src/server/vnparse.c

```diff
@@ -3104,7 +3104,98 @@ relnodes_input_select_init(relnodes_input_select_t *r_input)
 	r_input->good_vnodes = NULL;
 }
 
-/*
+/**
+ *
+ * @brief
+ *	Return a subset of the exec_vnode string 'e_vnode'
+ *	where only vnode chunks matching the vnode names
+ *	in 'vnl_good' (list of healthy vnodes) are shown.
+ * @par
+ *	For example, given:
+ *		e_vnode = (vn1:<r1>=<v1>)+(vn2:<r2>=<v2>)+(vn3:<r3>=<v3>)
+ *	and
+ *		vnl_fails lists: vn2
+ *
+ *	then this function returns:
+ *		e_vnode = (vn1:<r1>=<v1>)+(vn3:<r3>=<v3>)
+ *
+ * @param[in]	e_vnode - exec_vnode string
+ * @param[in]	vnl_fails - some list of names of failed vnodes
+ *
+ * @return char *
+ * @retval != NULL	the exec_vnode string of healthy vnodes
+ * @retval == NULL	if error encountered.
+ *
+ * @note
+ *	The returned string points to a statically allocated buffer
+ *	that must not be freed, and will get overwritten on the
+ *	next call to this function.
+ *
+ */
+static char *
+return_available_vnodes(char *e_vnode, vnl_t *vnl_good)
+{
+	char		*save_ptr;	/* posn for strtok_r() */
+	char		*pc;
+	char		*tmpbuf;
+	static	char	*ebuf = NULL;
+	static int	ebuf_size = 0;
+
+	if (e_vnode == NULL)
+		return NULL;
+
+	tmpbuf = strdup(e_vnode);
+	if (tmpbuf == NULL)
+		return NULL;
+
+	if (ebuf == NULL) {
+		ebuf = malloc(LOG_BUF_SIZE + 1);
+		if (ebuf == NULL) {
+			log_err(errno, __func__, "malloc failed");
+			return NULL;
```

Comment:
Make sure you're freeing tmpbuf before you return from this function. (multiple cases)

Reply:
Yes, I've taken care of this.